glean-core/csharp/Glean/Net/BaseUploader.cs

```diff
@@ -0,0 +1,154 @@
+            }
+
+            if (config.pingTag != null)
+            {
+                headerList.Add(("X-Debug-ID", config.pingTag));
+            }
+
+            return headerList.ToArray();
+        }
+
+        /// <summary>
+        /// Signals Glean to upload pings at the next best opportunity.
+        /// </summary>
+        internal void TriggerUpload(Configuration config)
+        {
+            // TODO: must not work like this, it should work off the main thread.
+            // FOR TESTING Implement the upload worker here and call this from Glean.cs
+
+            int waitAttempts = 0;
+            int uploadFailures = 0;
+
+            while (uploadFailures < 3)
+            {
+                FfiUploadTask incomingTask = new FfiUploadTask();
+                LibGleanFFI.glean_get_upload_task(ref incomingTask, config.logPings);
+
+                UploadTaskTag tag = (UploadTaskTag)incomingTask.tag;
+                switch (tag)
+                {
+                    case UploadTaskTag.Upload:
+                        {
+                            // Extract C#-friendly data from the FFI object
+                            string documentId = LibGleanFFI.GetFromRustString(incomingTask.body.documentId);
+                            string path = LibGleanFFI.GetFromRustString(incomingTask.body.path);
+                            string headersString = LibGleanFFI.GetFromRustString(incomingTask.body.headers);
+                            (string, string)[] headers = GetHeadersFromJSONString(documentId, headersString, config);
+                            byte[] body = new byte[incomingTask.body.bodyLen];
+                            Marshal.Copy(incomingTask.body.body, body, 0, body.Length);
+
+                            // Delegate the actual upload and get its return value.
+                            UploadResult result = Upload(path, body, headers, config);
+
+                            if (result is RecoverableFailure)
+                            {
+                                uploadFailures += 1;
+                            }
+
+                            // Copy the `FfiUploadTask` to unmanaged memory, because
+                            // `glean_process_ping_upload` assumes it has to free the memory.
+                            IntPtr ptrCopy = Marshal.AllocHGlobal(Marshal.SizeOf(incomingTask));
+                            Marshal.StructureToPtr(incomingTask, ptrCopy, false);
+
+                            // Process the upload response in the core.
+                            LibGleanFFI.glean_process_ping_upload_response(ptrCopy, result.ToFfi());
+
+                            // Free the allocated.
+                            Marshal.FreeHGlobal(ptrCopy);
+                        }
+                        break;
+                    case UploadTaskTag.Wait:
+                        {
+                            if (waitAttempts < 3)
```

Comment:
is it `maxRetries` ?

Reply:
Nope, this is really wait attempts: https://searchfox.org/glean/rev/73522afaaa64a2c904ec28e209e53c92fb94967b/glean-core/python/glean/net/ping_upload_worker.py#154 - we don't want to get into any infinite loop if there's no connectivity!